File: coding/src/com/algos/li05_binarysearch/L153_FindMinimumInRotatedSortedArray.java
package  com.algos.li05_binarysearch;

public class L153_FindMinimumInRotatedSortedArray {

	public static void main(String[] args) {
		L153_FindMinimumInRotatedSortedArray obj = new L153_FindMinimumInRotatedSortedArray();
		System.out.println(obj.findMin(new int[] {4,5,6,7,0,1,2}));
	}
	
    // Problem: https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
    // Idea: Continuosly, update min with nums[mid]
	public int findMin(int[] nums) {
        int left = 0, right = nums.length-1;
        int min = nums[0];
        while(left<=right){
            if(nums[left] <= nums[right]){		//[2,1]
            	min = Math.min(min, nums[left]);
                break;
            }
            
            int mid = (left+right)/2;
            min = Math.min(min, nums[mid]);		//[3,1,2]
            if(nums[left] <= nums[mid]){
                left = mid+1;
            }else{
                right = mid-1;
            }
        }
        
        return min;
    }

}

File: dsa/patterns/05_binarysearch/python/l153_find_minimum_in_rotated_sorted_array.py
# LeetCode 153 - Find Minimum in Rotated Sorted Array
#
# Idea:
# If the current subarray is already sorted, the minimum is at the left.
# Otherwise:
# - If left half is sorted, minimum is nums[low], discard left half
# - Else right half is sorted, minimum is nums[mid], discard right half
#
# Time Complexity: O(log n)
# Space Complexity: O(1)

def find_min(nums: list[int]) -> int:
    left, right = 0, len(nums) - 1
    minnum = float("inf")

    while left <= right:
        mid = left + (right - left) // 2

        if nums[left] <= nums[mid]:
            minnum = min(minnum, nums[left])
            left = mid + 1
        else:
            minnum = min(minnum, nums[mid])
            right = mid - 1
    return minnum


# Driver code
if __name__ == "__main__":
    nums = [4, 5, 6, 7, 0, 1, 2]
    print(find_min(nums))  # 0
